File: clients/python/neuroindex_client.py
#!/usr/bin/env python3
"""
NeuroIndex Python Client
Connects to NeuroIndex RESP protocol server
"""

import socket
import sys


class NeuroIndexClient:
    """Python client for NeuroIndex database using RESP protocol"""

    def __init__(self, host='localhost', port=6381):
        self.host = host
        self.port = port
        self.socket = None
        self.connect()

    def connect(self):
        """Connect to the NeuroIndex server"""
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.connect((self.host, self.port))
            print(f"✓ Connected to NeuroIndex at {self.host}:{self.port}")
        except Exception as e:
            print(f"✗ Failed to connect: {e}")
            sys.exit(1)

    def close(self):
        """Close the connection"""
        if self.socket:
            self.socket.close()
            print("✓ Connection closed")

    def _send_command(self, *args):
        """Send a RESP command to the server"""
        # Build RESP array
        cmd = f"*{len(args)}\r\n"
        for arg in args:
            arg_str = str(arg)
            cmd += f"${len(arg_str)}\r\n{arg_str}\r\n"

        self.socket.sendall(cmd.encode())

    def _read_response(self):
        """Read and parse RESP response from server"""
        line = self._read_line()

        if not line:
            return None

        resp_type = line[0]
        data = line[1:]

        if resp_type == '+':  # Simple string
            return data
        elif resp_type == '-':  # Error
            raise Exception(f"Server error: {data}")
        elif resp_type == ':':  # Integer
            return int(data)
        elif resp_type == '$':  # Bulk string
            length = int(data)
            if length == -1:
                return None
            value = self.socket.recv(length).decode()
            self._read_line()  # Read trailing \r\n
            return value
        elif resp_type == '*':  # Array
            count = int(data)
            if count == -1:
                return None
            return [self._read_response() for _ in range(count)]
        else:
            raise Exception(f"Unknown RESP type: {resp_type}")

    def _read_line(self):
        """Read a line from the socket (until \\r\\n)"""
        line = b''
        while True:
            char = self.socket.recv(1)
            if not char:
                return None
            if char == b'\r':
                self.socket.recv(1)  # Read \n
                break
            line += char
        return line.decode()

    # Command methods

    def ping(self, message=None):
        """Test connection"""
        if message:
            self._send_command("PING", message)
        else:
            self._send_command("PING")
        return self._read_response()

    def echo(self, message):
        """Echo a message"""
        self._send_command("ECHO", message)
        return self._read_response()

    def set(self, key, value):
        """Set a key-value pair"""
        self._send_command("SET", key, value)
        return self._read_response()

    def get(self, key):
        """Get a value by key"""
        self._send_command("GET", key)
        return self._read_response()

    def delete(self, *keys):
        """Delete one or more keys"""
        self._send_command("DEL", *keys)
        return self._read_response()

    def exists(self, *keys):
        """Check if keys exist"""
        self._send_command("EXISTS", *keys)
        return self._read_response()

    def mget(self, *keys):
        """Get multiple values"""
        self._send_command("MGET", *keys)
        return self._read_response()

    def mset(self, *pairs):
        """Set multiple key-value pairs. Usage: mset('key1', 'val1', 'key2', 'val2')"""
        if len(pairs) % 2 != 0:
            raise ValueError("MSET requires an even number of arguments")
        self._send_command("MSET", *pairs)
        return self._read_response()

    def keys(self, pattern="*"):
        """Get all keys (pattern matching not implemented)"""
        self._send_command("KEYS", pattern)
        return self._read_response()

    def dbsize(self):
        """Get database size"""
        self._send_command("DBSIZE")
        return self._read_response()

    def info(self):
        """Get server info"""
        self._send_command("INFO")
        return self._read_response()


def main():
    """Example usage"""
    print("NeuroIndex Python Client Demo")
    print("=" * 50)

    # Connect
    client = NeuroIndexClient(host='localhost', port=6381)

    try:
        # Test ping
        print(f"\nPING: {client.ping()}")
        print(f"PING with message: {client.ping('Hello NeuroIndex!')}")

        # Set and get
        print(f"\nSET user:1 'Alice': {client.set('user:1', 'Alice')}")
        print(f"GET user:1: {client.get('user:1')}")

        # Multiple operations
        print(f"\nMSET: {client.mset('user:2', 'Bob', 'user:3', 'Charlie', 'user:4', 'David')}")
        print(f"MGET: {client.mget('user:1', 'user:2', 'user:3', 'user:4')}")

        # Exists and delete
        print(f"\nEXISTS user:1 user:5: {client.exists('user:1', 'user:5')}")
        print(f"DEL user:1: {client.delete('user:1')}")
        print(f"EXISTS user:1: {client.exists('user:1')}")

        # Database info
        print(f"\nDBSIZE: {client.dbsize()}")
        print(f"KEYS: {client.keys()}")

        # Server info
        print(f"\nINFO:\n{client.info()}")

        print("\n" + "=" * 50)
        print("✓ All operations completed successfully!")

    except Exception as e:
        print(f"\n✗ Error: {e}")
    finally:
        client.close()


if __name__ == "__main__":
    main()

File: clients/go/neuroindex_client.go
// NeuroIndex Go Client
// Connects to NeuroIndex RESP protocol server

package main

import (
	"bufio"
	"errors"
	"fmt"
	"net"
	"strconv"
	"strings"
)

// NeuroIndexClient represents a connection to NeuroIndex server
type NeuroIndexClient struct {
	host   string
	port   int
	conn   net.Conn
	reader *bufio.Reader
}

// NewClient creates a new NeuroIndex client
func NewClient(host string, port int) (*NeuroIndexClient, error) {
	client := &NeuroIndexClient{
		host: host,
		port: port,
	}
	err := client.Connect()
	if err != nil {
		return nil, err
	}
	return client, nil
}

// Connect establishes connection to the server
func (c *NeuroIndexClient) Connect() error {
	addr := fmt.Sprintf("%s:%d", c.host, c.port)
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return fmt.Errorf("failed to connect: %w", err)
	}

	c.conn = conn
	c.reader = bufio.NewReader(conn)
	fmt.Printf("✓ Connected to NeuroIndex at %s\n", addr)
	return nil
}

// Close closes the connection
func (c *NeuroIndexClient) Close() error {
	if c.conn != nil {
		fmt.Println("✓ Connection closed")
		return c.conn.Close()
	}
	return nil
}

// sendCommand sends a RESP command to the server
func (c *NeuroIndexClient) sendCommand(args ...string) error {
	// Build RESP array
	cmd := fmt.Sprintf("*%d\r\n", len(args))
	for _, arg := range args {
		cmd += fmt.Sprintf("$%d\r\n%s\r\n", len(arg), arg)
	}

	_, err := c.conn.Write([]byte(cmd))
	return err
}

// readResponse reads and parses RESP response from server
func (c *NeuroIndexClient) readResponse() (interface{}, error) {
	line, err := c.reader.ReadString('\n')
	if err != nil {
		return nil, err
	}

	line = strings.TrimRight(line, "\r\n")
	if len(line) == 0 {
		return nil, errors.New("empty response")
	}

	respType := line[0]
	data := line[1:]

	switch respType {
	case '+': // Simple string
		return data, nil

	case '-': // Error
		return nil, fmt.Errorf("server error: %s", data)

	case ':': // Integer
		return strconv.ParseInt(data, 10, 64)

	case '$': // Bulk string
		length, err := strconv.Atoi(data)
		if err != nil {
			return nil, err
		}
		if length == -1 {
			return nil, nil
		}

		buf := make([]byte, length+2) // +2 for \r\n
		_, err = c.reader.Read(buf)
		if err != nil {
			return nil, err
		}
		return string(buf[:length]), nil

	case '*': // Array
		count, err := strconv.Atoi(data)
		if err != nil {
			return nil, err
		}
		if count == -1 {
			return nil, nil
		}

		array := make([]interface{}, count)
		for i := 0; i < count; i++ {
			element, err := c.readResponse()
			if err != nil {
				return nil, err
			}
			array[i] = element
		}
		return array, nil

	default:
		return nil, fmt.Errorf("unknown RESP type: %c", respType)
	}
}

// Command methods

// Ping tests the connection
func (c *NeuroIndexClient) Ping(message ...string) (string, error) {
	if len(message) > 0 {
		err := c.sendCommand("PING", message[0])
		if err != nil {
			return "", err
		}
	} else {
		err := c.sendCommand("PING")
		if err != nil {
			return "", err
		}
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	return resp.(string), nil
}

// Echo echoes a message
func (c *NeuroIndexClient) Echo(message string) (string, error) {
	err := c.sendCommand("ECHO", message)
	if err != nil {
		return "", err
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	return resp.(string), nil
}

// Set sets a key-value pair
func (c *NeuroIndexClient) Set(key, value string) (string, error) {
	err := c.sendCommand("SET", key, value)
	if err != nil {
		return "", err
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	return resp.(string), nil
}

// Get gets a value by key
func (c *NeuroIndexClient) Get(key string) (string, error) {
	err := c.sendCommand("GET", key)
	if err != nil {
		return "", err
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	if resp == nil {
		return "", nil
	}
	return resp.(string), nil
}

// Del deletes one or more keys
func (c *NeuroIndexClient) Del(keys ...string) (int64, error) {
	args := append([]string{"DEL"}, keys...)
	err := c.sendCommand(args...)
	if err != nil {
		return 0, err
	}

	resp, err := c.readResponse()
	if err != nil {
		return 0, err
	}
	return resp.(int64), nil
}

// Exists checks if keys exist
func (c *NeuroIndexClient) Exists(keys ...string) (int64, error) {
	args := append([]string{"EXISTS"}, keys...)
	err := c.sendCommand(args...)
	if err != nil {
		return 0, err
	}

	resp, err := c.readResponse()
	if err != nil {
		return 0, err
	}
	return resp.(int64), nil
}

// MGet gets multiple values
func (c *NeuroIndexClient) MGet(keys ...string) ([]interface{}, error) {
	args := append([]string{"MGET"}, keys...)
	err := c.sendCommand(args...)
	if err != nil {
		return nil, err
	}

	resp, err := c.readResponse()
	if err != nil {
		return nil, err
	}
	return resp.([]interface{}), nil
}

// MSet sets multiple key-value pairs
func (c *NeuroIndexClient) MSet(pairs ...string) (string, error) {
	if len(pairs)%2 != 0 {
		return "", errors.New("MSET requires an even number of arguments")
	}

	args := append([]string{"MSET"}, pairs...)
	err := c.sendCommand(args...)
	if err != nil {
		return "", err
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	return resp.(string), nil
}

// Keys gets all keys
func (c *NeuroIndexClient) Keys(pattern ...string) ([]interface{}, error) {
	p := "*"
	if len(pattern) > 0 {
		p = pattern[0]
	}

	err := c.sendCommand("KEYS", p)
	if err != nil {
		return nil, err
	}

	resp, err := c.readResponse()
	if err != nil {
		return nil, err
	}
	return resp.([]interface{}), nil
}

// DBSize gets database size
func (c *NeuroIndexClient) DBSize() (int64, error) {
	err := c.sendCommand("DBSIZE")
	if err != nil {
		return 0, err
	}

	resp, err := c.readResponse()
	if err != nil {
		return 0, err
	}
	return resp.(int64), nil
}

// Info gets server info
func (c *NeuroIndexClient) Info() (string, error) {
	err := c.sendCommand("INFO")
	if err != nil {
		return "", err
	}

	resp, err := c.readResponse()
	if err != nil {
		return "", err
	}
	return resp.(string), nil
}

// Example usage
func main() {
	fmt.Println("NeuroIndex Go Client Demo")
	fmt.Println(strings.Repeat("=", 50))

	// Connect
	client, err := NewClient("localhost", 6381)
	if err != nil {
		fmt.Printf("✗ Failed to connect: %v\n", err)
		return
	}
	defer client.Close()

	// Test ping
	pong, _ := client.Ping()
	fmt.Printf("\nPING: %s\n", pong)

	pongMsg, _ := client.Ping("Hello NeuroIndex!")
	fmt.Printf("PING with message: %s\n", pongMsg)

	// Set and get
	ok, _ := client.Set("user:1", "Alice")
	fmt.Printf("\nSET user:1 'Alice': %s\n", ok)

	value, _ := client.Get("user:1")
	fmt.Printf("GET user:1: %s\n", value)

	// Multiple operations
	ok, _ = client.MSet("user:2", "Bob", "user:3", "Charlie", "user:4", "David")
	fmt.Printf("\nMSET: %s\n", ok)

	values, _ := client.MGet("user:1", "user:2", "user:3", "user:4")
	fmt.Printf("MGET: %v\n", values)

	// Exists and delete
	count, _ := client.Exists("user:1", "user:5")
	fmt.Printf("\nEXISTS user:1 user:5: %d\n", count)

	deleted, _ := client.Del("user:1")
	fmt.Printf("DEL user:1: %d\n", deleted)

	count, _ = client.Exists("user:1")
	fmt.Printf("EXISTS user:1: %d\n", count)

	// Database info
	size, _ := client.DBSize()
	fmt.Printf("\nDBSIZE: %d\n", size)

	keys, _ := client.Keys()
	fmt.Printf("KEYS: %v\n", keys)

	// Server info
	info, _ := client.Info()
	fmt.Printf("\nINFO:\n%s\n", info)

	fmt.Println()
	fmt.Println(strings.Repeat("=", 50))
	fmt.Println("✓ All operations completed successfully!")
}
